### ISSUE TITLE
determinism, seeding and caching

### ISSUE THREAD

**Issue by michellemay** [DiUS/java-faker#326]
Hi, I'm wondering, is there a way to achieving determinism?
I can't get coherent a person data with current methods unless I carefully make sure the calling order is fixed, re-seeding before each calls and make sure I do not call composite methods.

For instance, I want to have a coherent username, first name, last name and email values.  If I call faker.name().lastName first, there is no way I can get a coherent faker.name().username() using the current api even if I re-seed the random number generator.  The reason is because username() will call firstName() first and use the current state of the PRNG and modify it before lastName is called..

A possible solution would be to have a cache in the service and reuse cached values unless the cache is cleared.  What do you think ?  Maybe I missed something along the way.  If this is the case, please accept my apologies.

**Comment by michellemay** [comment]
I ended up subclassing Faker and FakeValuesService and overriding some methods.  This looks more like a hack than anything else.   It could have been cleaner given less strict members/attributes visibility.

**Comment by UngarMax** [comment]
@michellemay Could you please explain how did you do it?

**Comment by michellemay** [comment]
It's not perfect but it does the job for my needs.  So I did the following:

1) Subclass `FakeValuesService`:
 - add a mutable hashmap cache.
 - override `resolve` and `expression` and lookup in the cache before delegating to the base class call.

2) Subclass `Faker`
 - embed an instance of `FakeValuesServiceWithCache`
 - add method clearCache that clear the cache
 - override `fakeValuesService`, `resolve` and `expression` to call the embedded service instance

I clear the faker cache between each records.  I also I keep a mapping table between original and faked data in the calling code to resolve with the same value 

It works because we have all dependent values from the same record.  This way, we can have coherent fullname and email for instance. 

It could probably have been implemented at a different level but I did not want to fork the repo and maintain that.
